File: bindings/python/test/Library/Mathematics/Geometry/D3/Transformations/Rotations/Quaternion.py
# coding=utf-8

################################################################################################################################################################

# @project        Library ▸ Mathematics
# @file           LibraryMathematicsPy/Geometry/D3/Transformations/Rotations/Geometry.py
# @author         Lucas Brémond <lucas@loftorbital.com>
# @license        Apache License 2.0

################################################################################################################################################################

import numpy

from Library.Core import Types

from LibraryMathematicsPy import Geometry

Angle = Geometry.Angle
Quaternion = Geometry.D3.Transformations.Rotations.Quaternion
RotationVector = Geometry.D3.Transformations.Rotations.RotationVector

################################################################################################################################################################

q = Quaternion(0.0, 0.0, 0.0, 1.0, Quaternion.Format.XYZS)

assert (q == q) is True
assert (q != q) is False

print('q:', q)

assert q.isDefined() is True
assert q.isUnitary() is True
assert q.isNear(q, Angle.Degrees(0.0)) is True

assert q.x() == 0.0
assert q.y() == 0.0
assert q.z() == 0.0
assert q.s() == 1.0
assert numpy.array_equal(q.getVectorPart(), numpy.array([[0.0], [0.0], [0.0]], dtype=float))
assert q.getScalarPart() == 1.0
assert q.toNormalized() == q
assert q.toConjugate() == q
assert q.toInverse() == q
# print('pow:', q.pow())
# print('exp:', q.exp())
# print('log:', q.log())
assert q.norm() == 1.0
assert q.crossMultiply(q) == q
assert q.dotMultiply(q) == q
assert numpy.array_equal(q.rotateVector(numpy.array([[0.0], [0.0], [1.0]], dtype=float)), numpy.array([[0.0], [0.0], [1.0]], dtype=float))
assert numpy.array_equal(q.toVector(Quaternion.Format.XYZS), numpy.array([[0.0], [0.0], [0.0], [1.0]], dtype=float)) is True
assert q.toString(Quaternion.Format.XYZS) == "[0.0, 0.0, 0.0, 1.0]"

q.normalize()

assert q == q

q.conjugate()

assert q.toString(Quaternion.Format.XYZS) == "[-0.0, -0.0, -0.0, 1.0]"

q.inverse()

assert q.toString(Quaternion.Format.XYZS) == "[0.0, 0.0, 0.0, 1.0]"

q.rectify()

assert q.toString(Quaternion.Format.XYZS) == "[0.0, 0.0, 0.0, 1.0]"

assert Quaternion.Undefined().isDefined() is False
assert Quaternion.Unit().isDefined() is True
assert Quaternion.XYZS(0.0, 0.0, 0.0, 1.0).isDefined() is True
assert Quaternion.RotationVector(RotationVector(numpy.array([[0.0], [0.0], [1.0]], dtype=float), Angle.Zero())).isDefined() is True
# RotationMatrix
assert Quaternion.Parse("[0.0, 0.0, 0.0, 1.0]", Quaternion.Format.XYZS).isDefined() is True

################################################################################################################################################################

File: bindings/python/src/LibraryMathematicsPy/Geometry/3D/Transformations/Rotations/Quaternion.cpp
////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

/// @project        Open Space Toolkit ▸ Mathematics
/// @file           LibraryMathematicsPy/Geometry/3D/Transformations/Rotations/Quaternion.cpp
/// @author         Lucas Brémond <lucas@loftorbital.com>
/// @license        Apache License 2.0

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

#include <LibraryMathematicsPy/Utilities/IterableConverter.hpp>

#include <Library/Mathematics/Geometry/3D/Transformations/Rotations/Quaternion.hpp>
#include <Library/Mathematics/Geometry/3D/Transformations/Rotations/RotationVector.hpp>
#include <Library/Mathematics/Geometry/3D/Transformations/Rotations/RotationMatrix.hpp>

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

inline void                     LibraryMathematicsPy_Geometry_3D_Transformations_Rotations_Quaternion ( )
{

    using namespace boost::python ;

    using library::core::types::Real ;
    using library::core::types::String ;

    using library::math::obj::Vector3d ;
    using library::math::obj::Vector4d ;
    using library::math::geom::d3::trf::rot::Quaternion ;

    scope in_Quaternion = class_<Quaternion>("Quaternion", init<const Real&, const Real&, const Real&, const Real&, const Quaternion::Format&>())

        .def(init<const Vector4d&, const Quaternion::Format&>())
        .def(init<const Vector3d&, const Real&>())

        .def(self == self)
        .def(self != self)

		.def(self * self)
		.def(self * Vector3d())

		.def(self / self)

        .def(self_ns::str(self_ns::self))

        .def("__repr__", +[] (const Quaternion& aQuaternion) -> std::string { return aQuaternion.toString() ; })

        .def("isDefined", &Quaternion::isDefined)
        .def("isUnitary", &Quaternion::isUnitary)
        .def("isNear", &Quaternion::isNear)

        .def("x", &Quaternion::x)
        .def("y", &Quaternion::y)
        .def("z", &Quaternion::z)
        .def("s", &Quaternion::s)
        .def("getVectorPart", &Quaternion::getVectorPart)
        .def("getScalarPart", &Quaternion::getScalarPart)
        .def("toNormalized", &Quaternion::toNormalized)
        .def("toConjugate", &Quaternion::toConjugate)
        .def("toInverse", &Quaternion::toInverse)
        // .def("pow", &Quaternion::pow)
        // .def("exp", &Quaternion::exp)
        // .def("log", &Quaternion::log)
        .def("norm", &Quaternion::norm)
        .def("crossMultiply", &Quaternion::crossMultiply)
        .def("dotMultiply", &Quaternion::dotMultiply)
        .def("rotateVector", &Quaternion::rotateVector)
        .def("toVector", &Quaternion::toVector)
        .def("toString", +[] (const Quaternion& aQuaternion) -> String { return aQuaternion.toString() ; })
        .def("toString", +[] (const Quaternion& aQuaternion, const Quaternion::Format aFormat) -> String { return aQuaternion.toString(aFormat) ; })
        .def("normalize", +[] (Quaternion& aQuaternion) -> void { aQuaternion.normalize() ; })
        .def("conjugate", +[] (Quaternion& aQuaternion) -> void { aQuaternion.conjugate() ; })
        .def("inverse", +[] (Quaternion& aQuaternion) -> void { aQuaternion.inverse() ; })
        .def("rectify", +[] (Quaternion& aQuaternion) -> void { aQuaternion.rectify() ; })
        .def("angularDifferenceWith", &Quaternion::angularDifferenceWith)

        .def("Undefined", &Quaternion::Undefined).staticmethod("Undefined")
        .def("Unit", &Quaternion::Unit).staticmethod("Unit")
        .def("XYZS", &Quaternion::XYZS).staticmethod("XYZS")
        .def("RotationVector", &Quaternion::RotationVector).staticmethod("RotationVector")
        .def("RotationMatrix", &Quaternion::RotationMatrix).staticmethod("RotationMatrix")
        .def("Parse", &Quaternion::Parse).staticmethod("Parse")

    ;

    enum_<Quaternion::Format>("Format")

        .value("XYZS", Quaternion::Format::XYZS)
        .value("SXYZ", Quaternion::Format::SXYZ)

    ;

    using library::core::ctnr::Array ;

    IterableConverter()

        .from_python<Array<Quaternion>>()
        .to_python<Array<Quaternion>>()

    ;

}

////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
